Improve ``Structure`` (subclass) construction performance.

brettwooldridge opened the pull request:
Fixes java-native-access/jna#1625 issue by caching the field list (``fieldList``) and validation state (``validationMap``) in addition to the existing ``layoutInfo`` and ``fieldOrder`` caches.

Caching this additional state provides substantial performance improvements.

dbwiddis commented on src/com/sun/jna/Structure.java:
```diff
@@ -1015,22 +1017,26 @@ protected void sortFields(List<Field> fields, List<String> names) {
-                int modifiers = fields[i].getModifiers();
-                if (Modifier.isStatic(modifiers) || !Modifier.isPublic(modifiers)) {
-                    continue;
+        synchronized (fieldList) {
```
I'm wondering whether synchronization is necessary here.  This will calculate the same result even if done by multiple threads, which would be a one-time "extra calculation" whereas checking the lock here will occur on every access of the map.

I realize the existing (heavy) calculations were guarded by a lock so it's still better than it was before, but it seems once initialized, an unlocked lookup will be even faster?

brettwooldridge replied:
If you allow multiple threads into this block the ``fieldList`` may experience a ``ConcurrentModificationException``.

dbwiddis replied:
> it is never the case that `fieldList` as a construct can be known to be fully initialized.

I may be missing something, but `fieldList` is initialized with the class as a static member, and is a fully initialized `WeakHashMap`.  

I had a look at the [Javadoc for it](https://docs.oracle.com/javase/8/docs/api/java/util/WeakHashMap.html) and am now even more confused. Apparently even synchronizing on the map object itself doesn't prevent the value and/or key from disappearing out from under you:
> The behavior of the WeakHashMap class depends in part upon the actions of the garbage collector, so several familiar (though not required) Map invariants do not hold for this class. Because the garbage collector may discard keys at any time, a WeakHashMap may behave as though an unknown thread is silently removing entries. In particular, even if you synchronize on a WeakHashMap instance and invoke none of its mutator methods, it is possible for the size method to return smaller values over time, for the isEmpty method to return false and then true, for the containsKey method to return true and later false for a given key, for the get method to return a value for a given key but later return null, for the put method to return null and the remove method to return false for a key that previously appeared to be in the map, and for successive examinations of the key set, the value collection, and the entry set to yield successively smaller numbers of elements.

The javadoc suggests using `Collections.synchronizedMap`.

Given the other maps already have this behavior it's entirely possible (probable) that I'm just not getting a subtlety here of how things work.

dbwiddis replied:
> If you allow multiple threads into this block the `fieldList` may experience a `ConcurrentModificationException`.

It looks like this only happens with an iterator, but we are looking up using a hash key rather than iterating, and that exception type [is not documented](https://docs.oracle.com/javase/8/docs/api/java/util/Map.html#computeIfAbsent-K-java.util.function.Function-) for `computeIfAbsent()`.

To be clear, I'm not opposed to the way you have written it as it follows the pattern of the others, and it's clearly an improvement. I just think the synchronization doesn't help much, and figured I'd ask the question.

matthiasblaesing replied:
To me this looks sane. The synchronization is not necessary to save time, but to protect the map. When a `put` operation is invoked, this might result in a resize of the internal table and if that is not protected it might result in undefined behavior.

I have mixed feelings about the synchronized blocks here in combination with `computeIfAbsent`, using `Collections#synchronizedMap` might be more straight forward, but from my POV it is ok as-is.

brettwooldridge replied:
@matthiasblaesing If you look at the original code for ``fieldOrder()`` for example:
```java
List<String> list = fieldOrder.get(clazz);
if (list == null) {
   list = getFieldOrder();
   fieldOrder.put(clazz, list);
}
return list;
```
The internal implementaton of ``computeIfAbsent()`` is doing essentially the same thing:
```java
V v;
if ((v = get(key)) == null) {
    V newValue;
    if ((newValue = mappingFunction.apply(key)) != null) {
        put(key, newValue);
        return newValue;
    }
}
return v;
```
So the resulting implementation in JNA reads much clearer:
```java
return fieldOrder.computeIfAbsent(clazz, (c) -> getFieldOrder());
```
And to your point regarding the need for synchronization, it is spot on. If access to the map is not synchronized (in some way) then a re-hash operation would be vulnerable to corruption.

Having said all of that, let us recognize that the *access pattern* of the ``fieldOrder`` map is overwhelmingly read vs. write. In fact, over the life of the JVM in which millions of ``Structures`` may be instantiated, the "write" to the map of the value of ``getFieldOrder()`` *occurs only once* in comparison to the millions of subsequent "reads".

If we *really* want an efficient design, we should use a ``ReentrantReadWriteLock``. It complicates the code in comparison to the above one-liner, but the performance gain is large (possibly huge).

Quoting ChatGPT:

> To optimize for a read-heavy scenario and avoid the overhead of unnecessary synchronization in the WeakHashMap for reads, we can use a ReentrantReadWriteLock. This type of lock allows multiple threads to read the cached data simultaneously while allowing exclusive access for writes.

```java
lock.readLock().lock();
try {
    Object cachedValue = fieldOrder.get(clazz);
    if (cachedValue != null) {
        return cachedValue;  // Return the cached result if found
    }
} finally {
    lock.readLock().unlock();
}

// If not found, compute the value under the write lock
lock.writeLock().lock();
try {
    // Double-check if another thread has computed the value before we do
    return fieldOrder.computeIfAbsent(clazz, (c) -> computeFieldOrder());
} finally {
    lock.writeLock().unlock();
}
```

I will be happy to implement this change if you guys think it's worth it, personally I am convinced it is. If this is the case, we should probably extend this pattern to all four caches present and consolidate to a single ``ReentrantReadWriteLock`` (because very quickly read accesses approach 100%).

dbwiddis replied:
> I will be happy to implement this change if you guys think it's worth it, personally I am convinced it is.

Would love to see some benchmarks.  In my reading I came to a similar conclusion.

brettwooldridge replied:
@dbwiddis @matthiasblaesing While I don't have a benchmark, I do have anecdotal performance numbers from our internal profiling showing roughly 15% time consumed in the ``Structure`` constructor when executing a load that creates millions of objects (Linux ``epoll``-related structures in this case). Unfortunately, as this is commercial data I cannot share detailed traces or screenshots.

Having said that, the time appeared to be a roughly equal split between ``validateFields()`` and ``initializeFields()``. Nearly all of the time in ``validateFields()`` will be eliminated (in this particular workload), and roughly half of time time in ``initializeFields()`` will be eliminated due to the call to ``getFieldList()`` which will now be cached. The remainder of time consumed in ``initializeFields()`` is not affected, goes way further down the stack, and is certainly not attempted to be covered here.

In addition to computational time spent before this change there was a substantial amount of garbage thrown off by these routines (and the Java reflection methods), something to the tune of just over 100MB in a one-minute snapshot of this particular load. This puts additional memory pressure on the collector which as an effect that is difficult to qualify but undeniably affects GC throughput.

If this change has a 10% performance improvement I would consider that a big win. While ten percent may not sound impactful to many in the case of a hyper-scaler it can be many tens of thousands $$ per year or more.

matthiasblaesing commented:
@brettwooldridge thank you. Two requests:

- Do you have performance numbers for this?
- Please add an entry to `CHANGES.md` (refer to the existing entries for the target format).

brettwooldridge commented:
@matthiasblaesing I have refactored the change using ``ReentrantReadWriteLock`` to substantially decrease contention as this is a 99.9% read/0.1% write type access pattern. Additionally, I updated the CHANGES.md as requested. All tests are passing.

dbwiddis reviewed (approved):
Thanks for the extra effort here!   I agree Structures are one of the biggest opportunities for performance improvements; the improvements in releasing the backing `Memory` in a previous release were pretty significant (and I still think there's room to improve with a "cache" of  the common 4- or 8- byte allocations) it's good to see unnecessary redundant calculations elimitated as well.

matthiasblaesing commented:
Thank you again and sorry for the delay in merging. Lets get this in.
